src/typing/env_js.ml

```diff
@@ -293,49 +342,93 @@ let already_bound_error =
    (since multiple declarations - sometimes but not always erroneous -
    may appear in an AST)
  *)
-let bind_entry cx name entry =
-  let scope = peek_scope () in
-  match Scope.get_entry name scope with
 
+let bind_entry cx name entry reason =
+  (* lex scopes can only hold let/const bindings
+   * var scopes can hold all bindings
+   * var bindings can't shadow lex bindings
+   * type entries are hoisted to var scope *)
+  let rec find_scope = function
+    | [] -> assert_false "empty scope list"
+    | scope::scopes -> Scope.(
+      match scope.kind, entry with
```

Comment:
You're right that these functions should be merged. I will do that. The bit of logic that complicates this, is that we need to check as we walk the lex scopes that a var binding doesn't shadow a lexical anywhere between the declaration site and the eventual var scope.